File: languages/typescript/io-large-file-streaming/stream.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import readline from 'node:readline';

const path = process.argv[2];
const agg = new Map();
let total = 0;
const rl = readline.createInterface({ input: fs.createReadStream(path), crlfDelay: Infinity });
for await (const line of rl) {
  if (!line) continue;
  const [, category, value, weight, active] = line.split(',');
  if (!agg.has(category)) {
    agg.set(category, { count: 0, value_sum: 0, weight_sum: 0, active_count: 0 });
  }
  const bucket = agg.get(category);
  bucket.count += 1;
  bucket.value_sum += Number(value);
  bucket.weight_sum += Number(weight);
  bucket.active_count += Number(active);
  total += 1;
}
console.log(JSON.stringify({ total_records: total, categories: Object.fromEntries(agg.entries()) }));

File: languages/python/io-large-file-streaming/stream.py
#!/usr/bin/env python3
import json
import sys
from collections import defaultdict

path = sys.argv[1]
agg = defaultdict(lambda: {"count": 0, "value_sum": 0, "weight_sum": 0, "active_count": 0})
total = 0
with open(path, 'r') as f:
    for line in f:
        line = line.strip()
        if not line:
            continue
        _, category, value, weight, active = line.split(',')
        bucket = agg[category]
        bucket['count'] += 1
        bucket['value_sum'] += int(value)
        bucket['weight_sum'] += int(weight)
        bucket['active_count'] += int(active)
        total += 1
print(json.dumps({"total_records": total, "categories": dict(agg)}, sort_keys=True))
